### PR TITLE
Added simple reproducible docker_build rule for ubuntu

### PR DESCRIPTION
**Pull request by aaron-prindle**

### CODE REVIEW

**Comment by aaron-prindle** [comment]
I have added bash as the entrypoint.  Running the built image appears to work, output below:
```
$ docker run -it bazel/repro
ducible:ubuntu
root@19145f907b58:/# ls
bin  boot  dev  etc  home  lib  lib64  media  mnt  opt  proc  root  run  sbin  srv  sys  tmp  usr  var
root@19145f907b58:/# ls
bin  boot  dev  etc  home  lib  lib64  media  mnt  opt  proc  root  run  sbin  srv  sys  tmp  usr  var
root@19145f907b58:/# ps aux
USER        PID %CPU %MEM    VSZ   RSS TTY      STAT START   TIME COMMAND
root          1  0.4  0.0  18236  3260 ?        Ss   20:05   0:00 bash
root         13  0.0  0.0  34420  2864 ?        R+   20:05   0:00 ps aux
root@19145f907b58:/# exit
exit
```

**Comment by aaron-prindle** [comment]
Have templated structure_test concept in the PR but it is not working currently.  If I run the tests "manually" they work but I am getting the following error running structure_test:
```
$ bazel run :ubuntu_test           
WARNING: /usr/local/google/home/aprindle/.cache/bazel/_bazel_aprindle/94ea7a73245e96603bcb84b183083fc5/external/bazel_tools/t       
ools/build_defs/pkg/pkg.bzl:196:9: pkg_tar: renaming non-dict `files` attribute to `srcs`.                                          
INFO: Found 1 target...                                                                                                             
Target //reproducible:ubuntu_test up-to-date:                                                                                       
  bazel-bin/reproducible/ubuntu_test                                                                                                
INFO: Elapsed time: 0.196s, Critical Path: 0.00s                                                                                    
                                                                                                                                    
INFO: Running command line: bazel-bin/reproducible/ubuntu_test                                                                      
+ reproducible/ubuntu                                                                                                               
                                                                                                                                    
cat: reproducible/ubuntu.runfiles/debian_docker/reproducible/ubuntu-layer.tar.sha256: No such file or directory                     
cat: reproducible/ubuntu.runfiles/debian_docker/reproducible/ubuntu.config: No such file or directory                               
No JSON object could be decoded                                                                                                     
reproducible/ubuntu: line 121: TOTAL_DIFF_IDS[@]: unbound variable                                                                  
cat: reproducible/ubuntu.runfiles/debian_docker/reproducible/ubuntu-layer.tar.sha256: No such file or directory                     
Error response from daemon: open /usr/local/google/docker/tmp/docker-import-008956528/sha256:: no such file or directory            
cat: reproducible/ubuntu.runfiles/debian_docker/reproducible/ubuntu-layer.tar.sha256: No such file or directory                     
cp: cannot stat ‘reproducible/ubuntu.runfiles/debian_docker/reproducible/ubuntu.config’: No such file or directory                  
ERROR: Non-zero return code '1' from command: Process exited with status 1.            
```

**Comment by aaron-prindle** [comment]
structure_tests added correctly, should work for travis now.  Ready for review

**Comment by dlorenc** [comment]
I got this confused in the debian image, see https://github.com/GoogleCloudPlatform/debian-docker/pull/99 for the correct way to set path and command.

File: WORKSPACE [hunk 1]
```diff
@@ -78,3 +78,9 @@ git_repository(
     commit = "3d73b4fecbd18de77588ab5eef712d50f34f601e",
     remote = "https://github.com/GoogleCloudPlatform/runtimes-common.git",
 )
+
+http_file(
+    name = "ubuntu_tar_download",
+    url = "http://cdimage.ubuntu.com/ubuntu-base/releases/16.04/release/ubuntu-base-16.04-core-amd64.tar.gz",
```
Comment (dlorenc) [inline WORKSPACE]: I think this location is better: https://partner-images.canonical.com/core/xenial/current/

File: tests/ubuntu_test.yaml [hunk 2]
```diff
@@ -0,0 +1,7 @@
+schemaVersion: '2.0.0'
```
[the remark this answers is not in the document]
Comment (aaron-prindle) [reply tests/ubuntu_test.yaml]: done

File: reproducible/BUILD [hunk 3]
```diff
@@ -71,3 +71,19 @@ structure_test(
     config = "//tests:debian_test.yaml",
     image = ":debian10",
 )
+
+docker_build(
+    name = "ubuntu",
+    tars = ["@ubuntu_tar_download//file"],
+    env = {"PORT": "8080"},
```
[the remark this answers is not in the document]
Comment (aaron-prindle) [reply reproducible/BUILD]: done

File: WORKSPACE (hunk 1 above)
Comment (aaron-prindle) [reply WORKSPACE]: done

**Comment by aaron-prindle** [comment]
done: changed env/path to be inline with #99

File: tests/ubuntu_test.yaml [hunk 4]
```diff
@@ -0,0 +1,21 @@
+schemaVersion: '1.0.0'
```
Comment (nkubala) [inline tests/ubuntu_test.yaml]: this needs to be `2.0.0` I believe. the original schema version didn't support the `args` field.

File: tests/ubuntu_test.yaml [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+  args: ['dump']
+  expectedOutput: ['Acquire::Retries "3"']
+  name: 'apt-config'
+- name: 'path'
+  command: 'sh'
+  args: ['-c', 'echo $PATH']
+  expectedOutput: ['/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin']
+fileExistenceTests:
+- name: 'Root'
+  isDirectory: true
```
Comment (nkubala) [inline tests/ubuntu_test.yaml]: this field also is deprecated in the latest schema version, in favor of using a unix-style permissions string (optional)

**Comment by dlorenc** [comment]
Yeah, the schema version needs to be 2, but I need to get in the bazel version of structure tests before that will work.

Or it could be 1, and be written in that schema.

File: tests/ubuntu_test.yaml (hunk 4 above)
Comment (aaron-prindle) [reply tests/ubuntu_test.yaml]: Does 2.0.0 work inside of bazel?  @dlorenc had commented that the schema version might need to be v1.0.0 to work with bazel:
https://github.com/GoogleCloudPlatform/debian-docker/pull/98#discussion_r143586836

**Comment by dlorenc** [comment]
I have a PR open against runtimes common to make it work. Once that goes in you should be good.

File: tests/ubuntu_test.yaml (hunk 5 above)
Comment (aaron-prindle) [reply tests/ubuntu_test.yaml]: I think this has to be updated in the debian{#}_test as well

Comment (dlorenc) [reply tests/ubuntu_test.yaml]: The new bazel rule should support both 1 and 2. But yeah, we should move everything to 2.

File: tests/ubuntu_test.yaml (hunk 4 above)
Comment (aaron-prindle) [reply tests/ubuntu_test.yaml]: done

File: tests/ubuntu_test.yaml [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+fileContentTests:
+- name: 'Ubuntu Distro Check'
+  expectedContents: ['Ubuntu']
+  path: '/etc/os-release'
+fileExistenceTests:
+- name: 'Root'
+  path: '/'
+  shouldExist: true
+  permissions: '-rw-r--r--'
+  fileExistenceTests:
```
Comment (dlorenc) [inline tests/ubuntu_test.yaml]: typo?

File: tests/ubuntu_test.yaml [hunk 7]
```diff
@@ -0,0 +1,21 @@
+schemaVersion: '1.0.0'
+commandTests:
+- name: 'apt-get'
```
Comment (dlorenc) [inline tests/ubuntu_test.yaml]: You can drop the apt-get one.

Comment (aaron-prindle) [reply tests/ubuntu_test.yaml]: done

File: tests/ubuntu_test.yaml (hunk 6 above)
Comment (aaron-prindle) [reply tests/ubuntu_test.yaml]: yes, fixed thanks

**Comment by aaron-prindle** [comment]
The test is failing as it cannot find /etc/os-release.  This is strange though as if I run the image in docker I am able to see the file /etc/os-release
```
$ cat   /usr/local/google/home/aprindle/.cache/bazel/_bazel_aprindle/94ea7a73245e96603bcb84b183083fc5/execroot/debian_docker/bazel-out/local-fastbuild/testlogs/reproducible/ubuntu_test/test.log                                                                                            
exec ${PAGER:-/usr/bin/less} "$0" || exit 1                                                                                         
-----------------------------------------------------------------------------                                                       
+ reproducible/ubuntu                                                                                                               
Loaded image ID: sha256:43090ab88adcaf7291776e346dafb23f6467aa42ea3e5aecead1a610dd9b90d4                                            
Tagging 43090ab88adcaf7291776e346dafb23f6467aa42ea3e5aecead1a610dd9b90d4 as bazel/reproducible:ubuntu                               
++ pwd                                                                                                                              
+ ../runtimes_common/structure_tests/go_default_test -image bazel/reproducible:ubuntu /usr/local/google/home/aprindle/.cache/       
bazel/_bazel_aprindle/94ea7a73245e96603bcb84b183083fc5/bazel-sandbox/8780106472488050902/execroot/debian_docker/bazel-out/loc       
al-fastbuild/bin/reproducible/ubuntu_test.runfiles/debian_docker/tests/ubuntu_test.yaml                                             
Using driver docker                                                                                                                 
2017/10/10 23:05:05 Running tests for file /usr/local/google/home/aprindle/.cache/bazel/_bazel_aprindle/94ea7a73245e96603bcb8       
4b183083fc5/bazel-sandbox/8780106472488050902/execroot/debian_docker/bazel-out/local-fastbuild/bin/reproducible/ubuntu_test.r       
unfiles/debian_docker/tests/ubuntu_test.yaml                                                                                        
--- FAIL: TestAll (1.35s)                                                                                                           
    --- FAIL: TestAll/File_Content_Test:_Ubuntu_Distro_Check (0.10s)                                                                
        structure.go:120: Failed to open /etc/os-release. Error: File /etc/os-release not found in image                            
        utils.go:29: Expected string .*Ubuntu.* not found in file contents!                                                         
        structure_test.go:46: Total tests run: 3                                                                                    
FAIL           
```
```                                                                                                                     
$ docker run -it bazel/reproducible:ubuntu                                                                           
root@0a8a3f56bd2c:/# cat /etc/os-release                                                                                            
NAME="Ubuntu"                                                                                                                       
VERSION="16.04.3 LTS (Xenial Xerus)"                                                                                                
ID=ubuntu                                                                                                                           
ID_LIKE=debian                                                                                                                      
PRETTY_NAME="Ubuntu 16.04.3 LTS"                                                                                                    
VERSION_ID="16.04"                                                                                                                  
HOME_URL="http://www.ubuntu.com/"                                                                                                   
SUPPORT_URL="http://help.ubuntu.com/"                                                                                               
BUG_REPORT_URL="http://bugs.launchpad.net/ubuntu/"                                                                                  
VERSION_CODENAME=xenial                                                                                                             
UBUNTU_CODENAME=xenial                                                                                                              
root@0a8a3f56bd2c:/#                   
```

**Comment by aaron-prindle** [comment]
Unsure how to debug this inside of structure_test.  I have removed the /etc/os-release test for now and with that test removed it is passing.  /etc/os-release does exist in the image but somehow structure_test is not finding it.

**Comment by dlorenc** [comment]
LGTM, let's get this in and fix the test later.

**Pull request merged**